examples/chatbot/src/websocket.rs

```diff
@@ -0,0 +1,172 @@
+use std::sync::Arc;
+
+use tokio::sync::Mutex;
+
+use eyre::WrapErr;
+use tokio_tungstenite::tungstenite;
+use tracing::Instrument;
+use twitch_api::{
+    eventsub::{
+        self,
+        event::websocket::{EventsubWebsocketData, ReconnectPayload, SessionData, WelcomePayload},
+        Event,
+    },
+    types::{self},
+    HelixClient,
+};
+use twitch_oauth2::{TwitchToken, UserToken};
+
+pub struct ChatWebsocketClient {
+    /// The session id of the websocket connection
+    pub session_id: Option<String>,
+    /// The token used to authenticate with the Twitch API
+    pub token: Arc<Mutex<UserToken>>,
+    /// The client used to make requests to the Twitch API
+    pub client: HelixClient<'static, reqwest::Client>,
+    /// The url to use for websocket
+    pub connect_url: url::Url,
+    /// Chats to connect to.
+    pub chats: Vec<twitch_api::types::UserId>,
+}
+
+impl ChatWebsocketClient {
+    /// Connect to the websocket and return the stream
+    async fn connect(
+        &self,
+    ) -> Result<
+        tokio_tungstenite::WebSocketStream<
+            tokio_tungstenite::MaybeTlsStream<tokio::net::TcpStream>,
+        >,
+        eyre::Error,
+    > {
+        tracing::info!("connecting to twitch");
+        let config = tungstenite::protocol::WebSocketConfig::default();
+        let (socket, _) =
+            tokio_tungstenite::connect_async_with_config(&self.connect_url, Some(config), false)
+                .await
+                .wrap_err("Can't connect")?;
+
+        Ok(socket)
+    }
+
+    /// Run the websocket subscriber
+    #[tracing::instrument(name = "subscriber", skip_all, fields())]
+    pub async fn run<Fut>(
+        mut self,
+        mut event_fn: impl FnMut(Event, types::Timestamp) -> Fut,
+    ) -> Result<(), eyre::Report>
+    where
+        Fut: std::future::Future<Output = Result<(), eyre::Report>>,
+    {
+        // Establish the stream
+        let mut s = self
+            .connect()
+            .await
+            .context("when establishing connection")?;
+        // Loop over the stream, processing messages as they come in.
+        loop {
+            tokio::select!(
+            Some(msg) = futures::StreamExt::next(&mut s) => {
```

Comment:
There's only one case in this `select!()` - this could be a `while let Some(msg) = s.next().await`, right?